### PR TITLE
add support for border image

### LINKED ISSUE #7317: Enhance border to support border-image
### Is this something you or your team is interested in contributing a pull request towards?
yes

### Expected Behavior
Right now we are not able to have something like a gradient color passed into border. 

### Actual Behavior
we would want to be able to pass a gradient color for border. 
Screenshot 2024-09-17 at 3 45 09 PM

### Steps to Reproduce

1. 2. 3.

### Your Environment

- Grommet version:
- Browser Name and version:
- Operating System and version (desktop or mobile):

### PR DESCRIPTION
**Pull request by britt6612**
#### What does this PR do?
This PR adds support for `borderImage` 
#### Where should the reviewer start?
utilz/border.js
#### What testing has been done on this PR?
added a test and story
#### Do Jest tests follow these best practices?

- [ ] `screen` is used for querying.
- [ ] The correct query is used. (Refer to [this list of queries](https://testing-library.com/docs/queries/about/#priority))
- [ ] `asFragment()` is used for snapshot testing.

#### Any background context you want to provide?
need for DS system
#### What are the relevant issues?
closes #7317 
#### Do the grommet docs need to be updated?
yes
#### Should this PR be mentioned in the release notes?
yes
#### Is this change backwards compatible or is it a breaking change?
 backwards compatible

### CODE REVIEW

**Comment by taysea** [comment]
I wonder if this should be structured like `border={{ image: }}` similar to how `border={{ color: }}` translates to border-color CSS

File: src/js/utils/index.d.ts [hunk 1]
```diff
@@ -185,6 +191,7 @@ export type BorderType =
       side?: BoxSideType;
       size?: BoxSizeType;
       style?: BoxStyleType;
+      image?: string;
```
Comment (MikeKingdom) [inline src/js/utils/index.d.ts:194]: Is there a corresponding change to propTypes?

Comment (britt6612) [reply src/js/utils/index.d.ts:194]: yes changing the structure a little so I will update :)

File: src/js/utils/border.js [hunk 2]
```diff
@@ -99,6 +118,19 @@ export const borderStyle = (borderData, responsive, theme) => {
         styles.push(breakpointStyle(breakpoint, responsiveStyle));
       }
     }
+    if (data.image) {
```
Comment (MikeKingdom) [inline src/js/utils/border.js:121]: Looks like this makes the border image silently supported on Grid, Table and Tag also (although without the proptypes and typescript support?)

Comment (britt6612) [reply src/js/utils/border.js:121]: well Tag adds on top of Box wouldnt the `proptypes` and `typescript` be there from Box?

Comment (britt6612) [reply src/js/utils/border.js:121]: I can add support for Grid,Table etc and look where else we are importing `borderStyle`

Comment (britt6612) [reply src/js/utils/border.js:121]: I did not add `border` to propTypes within Table we didnt even have the `border` in there before, I know there are places in grommet where under the hood certain functionality works but we don't always have it in TS and PropTypes.

**Review by jcfilben — APPROVED** [review]
Looks good!

File: src/js/utils/index.d.ts [hunk 3]
```diff
@@ -193,6 +200,7 @@ export type BorderType =
       side?: BoxSideType;
       size?: BoxSizeType;
       style?: BoxStyleType;
+      image?: string | { source?: string; slice?: number };
```
Comment (taysea) [inline src/js/utils/index.d.ts:203]: Wondering about the straight "string" support. We don't tend to support CSS shorthands directly on Grommet props. For example, I couldn't do: <Box border="1px dashed red" />

Should we just support the object for now?

Comment (britt6612) [reply src/js/utils/index.d.ts:203]: yeah I can change that.. good point

Comment (taysea) [reply src/js/utils/index.d.ts:203]: @jcfilben What do you think?

Comment (jcfilben) [reply src/js/utils/index.d.ts:203]: Yeah I think we could just support object for now, @taysea your right we don't normally include shorthand

Comment (britt6612) [reply src/js/utils/index.d.ts]: done!

File: src/js/utils/index.d.ts [hunk 4]
```diff
@@ -185,6 +191,7 @@ export type BorderType =
       side?: BoxSideType;
       size?: BoxSizeType;
       style?: BoxStyleType;
+      image?: {};
```
Comment (taysea) [inline src/js/utils/index.d.ts:194]:
```suggestion
      image?: { source?: string, slice?: number };
```

File: src/js/utils/index.d.ts [hunk 5]
```diff
@@ -193,6 +200,7 @@ export type BorderType =
...
```
Comment (taysea) [inline src/js/utils/index.d.ts:203]:
```suggestion
      image?: { source?: string, slice?: number };
```

File: src/js/utils/border.js [hunk 6]
```diff
@@ -21,22 +24,38 @@ export const responsiveBorderStyle = (data, theme) => {
...
+    if (typeof data.image === 'object') {
+      const { slice, source } = data.image;
+      if (source) {
+        styles.push(`border-image-source: ${source};`);
+      }
+      if (slice) {
+        styles.push(`border-image-slice: ${slice};`);
+      }
+    } else {
+      styles.push(`border-image: ${data.image};`);
```
Comment (taysea) [inline src/js/utils/border.js:55]: Looks like this still is supporting string?

File: src/js/utils/border.js [hunk 7]
```diff
@@ -99,6 +118,19 @@ export const borderStyle = (borderData, responsive, theme) => {
...
+      if (typeof data.image === 'object') {
+        const { slice, source } = data.image;
+        if (source) {
+          styles.push(`border-image-source: ${source};`);
+        }
+        if (slice) {
+          styles.push(`border-image-slice: ${slice};`);
+        }
+      } else {
+        styles.push(`border-image: ${data.image};`);
```
Comment (taysea) [inline src/js/utils/border.js:131]: Looks like this still is supporting string?

File: src/js/components/Box/stories/Border.js [hunk 8]
```diff
@@ -94,6 +94,17 @@ export const BorderBox = () => (
       <Text>Multiple Border</Text>
       <Text>With Between</Text>
     </Box>
+    <Box
+      pad="small"
+      border={{
+        size: 'small',
+        style: 'solid',
+        color: 'transparent',
+        image: 'linear-gradient(#f6b73c, #4d9f0c) 10',
```
Comment (taysea) [inline src/js/components/Box/stories/Border.js:103]: Can we update this to represent the new support of just object?

**Review by taysea — APPROVED** [review]
LGTM!

**Pull request merged**